### ISSUE TITLE
MoonScript + PEGs

### ISSUE THREAD

**Issue by jashkenas** [jashkenas/coffeescript#1593]
I know that folks have wanted to try doing CoffeeScript with a PEG parser for a while now, both because of perceived simplicity, as well as the potential to expand the grammar at runtime. There's now good prior art on parsing a CoffeeScript-esque language with a PEG, thanks to MoonScript.

http://moonscript.org/reference/

Check this out:

https://github.com/leafo/moonscript/blob/master/moonscript/parse.lua

If anyone wants to take a stab at a PEG parser, this would be a great place to start.

**Comment by eelco** [comment]
For the record, I would still be interested in this. 

Our product lets people build projects using CoffeeScript, but because Redux is not a drop-in replacement it’s very hard for us to switch the engine of the airplane mid-flight (and we also using the AST internally). 

But, the parser error reporting in Redux is so much better and these are the most common errors our users run into (and are stumped by).

I’m curious to hear how how much work it would be to port the Redux parser, any idea @michaelficarra?

**Comment by carlsmith** [comment]
Just to add, it would be relatively simple to write Pratt (top down operator precedence) parser for CoffeeScript (or almost any language).

Handwritten parsers are normally pretty daunting, but it's easier to write a Pratt parser than to define a language in BNF.

Pratt preempted apathy for his algorithm, and despite Crockford doing a lot to popularise it in recent years (and explain it in plain English), it is still pretty much unknown. It should be taught in every compiler class though. It's been described as the most beautiful algorithm ever, and is definitely up there.

**Comment by GeoffreyBooth** [comment]
A PR for this would be welcomed.
